### DISCUSSION TITLE
MQTT Publish without callback?

### DISCUSSION THREAD

**Discussion by johnchienbronci** [aws/aws-iot-device-sdk-cpp-v2#629] [Q&A] [answered]
aws-sdk-cpp: 1.11.145

I use the publish function to send IoT events. However, when dealing with a large number of publish() calls, I encounter situations where some packetId values don't have corresponding callback messages (the packet_id is always greater than 0).

I have a limitation that when the number of pending pubacks reaches 100, I will wait for the receipt of one puback before continuing to invoke publish().

The total number of invocations of the publish() function is approximately 1000. (The total packet count will not exceed the size limit of 512KB/s.)

a payload message length = 35 ~ 45 B 

How should I resolve this issue?
Are there any other resource limitations?

**Comment by bretambrose** [comment, marked as answer]
IoT Core has a [100 publish per second per connection](https://docs.aws.amazon.com/general/latest/gr/iot-core.html#message-broker-limits) limit.  In MQTT311, there is no way to negatively [acknowledge](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/errata01/os/mqtt-v3.1.1-errata01-os-complete.html#_Toc442180856) a publish (ie reject) and so IoT Core just drops publishes that exceed that limit.  In that case, there will never be a response; setting an [operation timeout](https://github.com/awslabs/aws-crt-cpp/blob/main/include/aws/crt/mqtt/MqttClient.h#L318) will cause the SDK client to give up waiting and notify you of a failure.  

Note that keeping the inflight publishes to < 100 doesn't address this limit since a good connection to AWS will enable you to exceed the 100 publish-per-second while staying under the inflight count (what MQTT5 calls [ReceiveMaximum](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901083)).

To avoid this you have a couple of options:
1.  Self-regulate your publishes to keep them under 100 TPS.
1.  Set an operation timeout per the instructions above and accept the failures that happen. 
1.  Switch to the MQTT5 client and, if constructed from the [builder](https://github.com/awslabs/aws-crt-cpp/blob/main/include/aws/iot/Mqtt5Client.h#L123), by default it will apply flow control to outbound publishes to keep it under 100TPS.  Additionally, even if you somehow exceed 100TPS, MQTT5 supports [negative pubacks](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901124), which means your rejected publishes will get an immediate client-side notification rather than just "disappearing."

**Reply by johnchienbronci** [reply]
Thanks for your reply and I will try to use mqtt5 reference (https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/1ddf4ab749155abe195acab9f9505791a93b9b45/samples/mqtt5/mqtt5_shared_subscription/main.cpp#L20)

Does the message of negative puback appear in the callback of publish?  Need additional implementation?

**Reply by jmklix** [reply]
@github-actions proposed-answer

You can get it with the result callback for the publish. You can see this being done in the example you linked [here](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/1ddf4ab749155abe195acab9f9505791a93b9b45/samples/mqtt5/mqtt5_shared_subscription/main.cpp#L324-L338):

```cpp
            std::shared_ptr<Mqtt5::PubAckPacket> puback =
                std::dynamic_pointer_cast<Mqtt5::PubAckPacket>(result->getAck());
            if (puback->getReasonCode() == 0)
            {
                fprintf(stdout, "[%s]: Publish Succeed.\n", publisher->name.c_str());
            }
            else
            {
                fprintf(
                    stdout,
                    "[%s] Publish failed. PubACK reason code: %d : %s\n",
                    publisher->name.c_str(),
                    puback->getReasonCode(),
                    puback->getReasonString()->c_str());
            }
```
